### ISSUE TITLE
[feature request]recomputable operation annotation

### ISSUE THREAD

**Issue by biolee** [tensorflow/tensorflow#13160]
'Training Deep Nets with Sublinear Memory Cost' and 'Memory-Efficient Implementation of DenseNets' indicate that use drop intermediate feature map and recompute it if needed can save memory(while add computation burden),

so we need some mechanism to annotate some op's inputs  `recomputable` , and drop this input memory after op finish(set input's reference count to 0), when need this op again, recompute it.

```python
a = tf.get_varible(shape=[None,10])
b = tf.get_varible(shape=[None,10])
c = tf.get_varible(shape=[None,10])
d = a*b
d_recomputable = tf.recomputable(d)
e = d_recomputable+c
```
relate issue 
- https://github.com/tensorflow/tensorflow/issues/1934
- https://github.com/tensorflow/tensorflow/issues/12948

in short, normal reference add reference count, recomputable reference do not add reference count

**Comment by yaroslavvb** [comment]
Marking the node as recomputable is not enough to save memory. You also have to specify when the recomputation happens.

TensorFlow scheduling logic (in [executor.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/executor.cc)) schedules nodes as evaluate soon as they are ready, so TensorFlow will happily recompute the node right away and hold it in memory which won't save you any memory.

Fun example -- something like `functools.reduce(tf.matmul, [tf.random_uniform((k,k))]*n)` needs O(n) memory in TensorFlow because by the time the first `matmul` is done, it will eval all the `random_uniform` nodes.

You can implement this recomputation on client level by inserting control dependencies to force a memory efficient order (like [here](https://github.com/yaroslavvb/stuff/tree/master/linearize)), and duplicating nodes you want to be recomputed (using tf.contrib.graph_editor like [here](https://github.com/yaroslavvb/stuff/blob/master/simple_rewiring.ipynb) or rewriting the graph in C++ API like [here](https://github.com/tensorflow/tensorflow/blob/cd4c17e63125da5e0047c3bdd05b483e86cc6759/tensorflow/core/grappler/optimizers/memory_optimizer.cc#L44)).

I do agree that current methods are quite awkward

**Comment by aselle** [comment]
@zheng-xq, do you want to comment about this issue, as I believe you are investigating automated ways of balancing memory/recomputation tradeoffs in your team.

**Comment by zheng-xq** [comment]
The current best practice to manually trade-off recomputation and memory is through TF functions. Through that, you can override the gradient with a custom function.

I'll leave for zffchen78@ to comment on more details.

**Comment by yaroslavvb** [comment]
@biolee btw, here's an [example](https://github.com/yaroslavvb/stuff/blob/master/saving%20memory%20by%20using%20functions.ipynb) of saving memory using functions. A function seems to be treated by tf.gradients as a single computation block and values inside are automatically recomputable.

I'm a bit confused on whether we can rely on this behavior in our models using default `tf.Defun` or whether the framework may start looking reusing computation inside functions at any version.

Some places use `noinline=True` attribute but the tests say its broken

(https://github.com/tensorflow/tensorflow/blob/01daba61e3a5099c6ad6439fa47e30c71560f06b/tensorflow/compiler/tests/function_test.py#L106)

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by ahundt** [comment]
I think this issue is still valid.

@zffchen78 you were mentioned in https://github.com/tensorflow/tensorflow/issues/13160#issuecomment-331797968 but I think it didn't notify you.

**Comment by zheng-xq** [comment]
Since some research groups use tf.Defun for recomputable operations, adding @josh11b to answer the question from @yaroslavvb that whether they can reply on this behavior, or what framework level support is on the roadmap.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by zffchen78** [comment]
The forementioned mechanism should be working. The test here illustrates w/ vs. w/o marking the forward function as noinline=True. 

https://github.com/tensorflow/tensorflow/blob/25d275280dfb163674f81c7681c2c1d34545a155/tensorflow/python/framework/function_test.py#L1435

To measure the runtime's peak memory usage, one can use ops demonstrated by this test:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/memory_stats/python/kernel_tests/memory_stats_ops_test.py

To see the effect more pronounced, one can increase the vector size (16 to something much larger):
https://github.com/tensorflow/tensorflow/blob/25d275280dfb163674f81c7681c2c1d34545a155/tensorflow/python/framework/function_test.py#L1459

@yaroslavvb I can not predict future. But several production scale models I'm working on depends on noinline=True to be respected and if it were broken somehow, some production teams would complain badly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by drpngx** [comment]
Nice, closing since there is a solution.

CC @zffchen78
